### PR TITLE
Node size depending on amount of links. Add hovering

### PR DESCRIPTION
**Pull request by dobrovolsky**
# Change shape of graph and add hovers

Add obsidian like graph with hover and circled graph.

[Demo (be careful, white color)](https://imgur.com/a/nWyzjP4)

## Motivation

You should have an ability to discover your structure and relation quickly.

### CODE REVIEW

**Comment by ingalless** [comment]
@tchayen @dobrovolsky it's looking like it's reaching a point that it might be good to ship a variety of graph.js files that can be selected from a configuration option

I like the graph a lot, but @ianjsikes changes from the other day are also very cool! And, in a few days, someone else might make another cool graph

I'd love to be able to choose

Fortunately, @ianjsikes also introduced the change which means the graph is its own file!

Really great work on this though, love the graph!

**Comment by nixsee** [comment]
Great work. I wonder if the size difference could be a bit more dramatic though? As in, that central node which has many links could be perhaps even 2x larger?

**Comment by tchayen** [comment]
Absolutely. Modularity is getting crucial in order to make any meaningful changes now.

IMO at this point, your PR @dobrovolsky is too good to be ignored and at the same time too different from what it looks like now to be merged.

@ingalless Exactly as you are saying, the closest best thing to reach now is it allow users to pick another `graph.js` file. Do you think it would be better to start with predefined ones or to allow for picking arbitrary file to use at one's own risk?

@nixsee do you have some idea how importing features could look like? It sounds amazing but I am not sure how to get there (and can't remember any vscode extensions doing such thing).

So tl;dr:
Changes like this are very welcome, but need a bit of preparation in terms of configuration, settings etc.

**Comment by ingalless** [comment]
@tchayen personally I don't think it hurts for them to be preconfigured options

It means we can get this graph in quicker _and_  allow others to start contributing their graphs which would be great.

It's simpler. And it means foam can preconfigure the setting to choose a default graph for users

**Comment by tchayen** [comment]
@dobrovolsky would you mind doing the following changes?

- creating a directory `/graphs` (or something similar)
- putting the current `graph.js` as `default.js` (or if you have a better name) and the one you made (we can go with a name `obsidian-like.js` for example)
- creating a configuration variable that will show up as select in settings to choose _graph representation_ between `Default` and `Obsidian-like` (with more to come in future)

Let me know if it sounds ok or if you prefer to not do it (you would like someone else do it) or want me to assist you a bit in doing that.

**Comment by dobrovolsky** [comment]
Now you can easily change graph appearance

**Review by ingalless — APPROVED** [review]
Great work @dobrovolsky! This looks really good stuff.

**Comment by tchayen** [comment]
@dobrovolsky it looks great!

File: src/extension.ts [hunk 1]
```diff
@@ -181,10 +189,16 @@ async function getWebviewContent(
         vscode.Uri.file(path.join(context.extensionPath, "static", fileName))
       )
       .toString();
+  
+  // replace ${name} with given variable
```
Comment (tchayen) [inline src/extension.ts:193]: Is this comment up to date? To be honest I have no idea what `${name}` it refers to.

File: static/graphs/obsidian.js [hunk 2]
```diff
@@ -0,0 +1,277 @@
+const MINIMAL_NODE_SIZE = 8
```
Comment (tchayen) [inline static/graphs/obsidian.js:1]: Very minor thing, but could you run Prettier on this file? I am seeing inconsistency with usage of semicolons for example.

File: static/webview.html [hunk 3]
```diff
@@ -30,6 +30,11 @@
         cursor: pointer;
         fill: var(--vscode-textLink-foreground);
       }
+      
+      .inactive {
+        opacity: 0.1;
+        transition: all 0.15s ease-out;
+      }
```
Comment (tchayen) [inline static/webview.html:37]: Looks like ideally we will need to have not only separate JS files but also CSS ones.

We could perhaps go with deeper structure like:

```
/graphs
  /default
    index.js
    style.css
  /obsidian-like
    index.js
    style.css
```

But I think we will be fine without doing such change at this point. It can wait (or of course feel free if you prefer to do it now yourself).

File: static/graphs/obsidian.js [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+  node = node.data(nodesData, (d) => d.id);
+  node.exit().remove();
+  node = node
+    .enter()
+    .append("circle")
+    .attr("r", (d) => {
+      return nodeSize[d.id]
+    })
+    .on("click", onClick)
+    .on("mouseover", mouseover)
```
Comment (tchayen) [inline static/graphs/obsidian.js:235]: Another very minor thing, but I think we could use some more consistent naming here.

`click` gives us `onClick` while `mouseover` leads to `mouseover`. Personally I would go with `on[...]` plus camelCase for all, like `onClick`, `onMouseOver` etc. but it is just a matter of my personal preference so feel free to go with whatever consistent way you prefer.

**Review by tchayen — COMMENTED** [review]
Some minor suggestions from me. Mostly regarding code style updates I would love to see (but nothing blocking).

Thanks again for all the hard work you have put into it! Can't wait to have it merged and released.

File: static/graphs/obsidian.js (hunk 2 above)
Comment (ingalless) [reply static/graphs/obsidian.js]: I spotted this too. Is there a `.prettierrc` in the repo?

Comment (tchayen) [reply static/graphs/obsidian.js]: I think not. I used to have it in projects where I wanted to change some settings, but if I am not planning to... should it be just empty `.prettierrc` with `{}` as content?

**Comment by dobrovolsky** [comment]
@tchayen just letting you know changes you asked has been made

File: static/graphs/obsidian.js (hunk 2 above)
Comment (ingalless) [reply static/graphs/obsidian.js]: @tchayen maybe the defaults for either `1.x` or `2.x` prettier would be a good idea? https://prettier.io/docs/en/configuration.html#basic-configuration this seems a good place to start.

But empty could work.

To me it feels that at least putting _something_ will mean we don't get odd scenarios like when @ianjsikes saved over a file that I'd saved before with a different version of prettier.

**Comment by Am3ra** [comment]
Don't know if it's been mentioned before, but is it possible to also scale the font size of the node label?
